### DISCUSSION TITLE
Salesforce - Data Deployment

### DISCUSSION THREAD

**Discussion by Akash-Bangalkar** [FlexDeploy-Community/community#70] [Show and tell]
Salesforce data deployment refers to the process of transferring, updating, or migrating data within the Salesforce platform from one Salesforce org to one or multiple other Salesforce orgs. After testing and approving data changes in a Sandbox or Development Org, they can be moved to Test and UAT sandboxes, and ultimately to the production environment.

FlexDeploy plays a significant role in streamlining and orchestrating the entire data deployment process within Salesforce. For comprehensive insights into how FlexDeploy facilitates data deployment within the Salesforce environment, please refer to the [Salesforce Data Deployment Tutorial](https://flexagon.atlassian.net/l/cp/yX64411Q).

_The data deployment process using FlexDeploy involves the following sequential steps:_

### Setting Up Salesforce Environments and Establishing External ID for Data Migration

To initiate the data migration process and ensure the unique identification of records within an object, it is imperative to establish a designated column or field as an identifier. Ideally, this field should already exist in the object and be indexed for efficient retrieval. The use of External IDs further streamlines the data migration process, allowing for seamless record updates, deletions, or upserts.

In instances where there is no pre-existing indexed field, the following steps outline the creation of a custom field to serve as the identifier:

1. Populating unique data into a new field may pose a challenge. To streamline the process, consider initially creating an Auto Number field and populating the initial data into all existing rows. Subsequently, the data type can be updated to Text, facilitating the easier insertion of unique data into the field.

![ExternlaID](https://github.com/FlexDeploy-Community/community/assets/125540444/93290458-14f7-4435-a530-53c744627a51)

2. Following the configuration of the External ID in all pertinent objects within the Development Org, it is crucial to propagate this change to all associated partial and full sandboxes, as well as the Production Org. This step guarantees the synchronization of the External ID field across all environments, facilitating a cohesive and harmonized data migration process.

### Setup in FlexDeploy for Salesforce Data Deployment

**1. Create Salesforce Package-Based Project:**
Begin by creating a Salesforce package-based project in FlexDeploy. If you already have an existing project, you can utilize it. Refer to the [Salesforce Tutorial](https://flexagon.atlassian.net/wiki/spaces/FD65/pages/10125818018/Salesforce+Tutorial) in case you need guidance on creating a Salesforce project in FlexDeploy.

**2. Configure External IDs:**
Access Project Configuration → Properties and input the external IDs. If maintaining the same external ID field name for all relevant objects, enter the external ID directly. For different external ID fields, use key=value pairs separated by commas, where the key represents the object, and the value is the external field ID for that object. Use the key 'All' for a default setting for all other objects (e.g., All=FD_External_ID__c, Contact=Email).

![2  External ID FD](https://github.com/FlexDeploy-Community/community/assets/125540444/31a586f8-d8e9-42df-859d-ea01f9f8c9de)

**3. Package Creation (Data Template Setup):**
Navigate to the Package option and create a package in FlexDeploy as usual. You can combine metadata files and data files within the same package.

**4. List of Objects for Migration:**
Make a list of all the objects you would like to migrate.

**5. Create Data Object Template File:**
Within the Package Option, go to the Package Files section and click on **Create** button to generate a new Data object template file for migration. Start with the top-level object and repeat the process for all objects, adjusting sequencing if needed.

![3  Create Data File](https://github.com/FlexDeploy-Community/community/assets/125540444/c7893ef5-f96c-45bc-934b-27e574de841e)

**6. Select Operation Type:**
Choose the operation type based on your requirements - Insert for creating records, Upsert & Update for updating records and Export Only for CSV backup without deployment.

**7. Select Object and Columns:**
Select the object and columns you wish to migrate.

**8. Enter Filters:**
Optionally, enter filters such as Eg., Rating = 'Hot', CreatedDate >= LAST_N_DAYS:1.

**9. Select External ID:**
Choose the External ID, ensuring it matches the configuration in the project properties and is included in the Column Names selection. Select External ID specifically for Upsert or Update operations; for Insert operations, it can be skipped as it is not required.

**10. Set Retry Count:**
Specify a value for the Retry Count to handle failed records during the data migration process. This feature enables automatic retries for failed records until either success is achieved or a predefined maximum retry count is reached.

**11. Define File Path:**
The file path defaults based on the selected object, but you can customize it for appropriateness. It needs to be unique in your project.

![4  Data Template](https://github.com/FlexDeploy-Community/community/assets/125540444/463128e9-bc2b-45eb-8086-fe4d0e31a6c9)

**12. Save the Configuration:**
Click on Save to finalize the configuration.

That completes the setup process. The package is now ready for build and deployment whenever needed. You can repeat the build to incorporate new data updates for the same set of files and fields.

### Build and Deploy

**1. Initiate Build Request:**
In Package Option, navigate to the package where you've created the file. Now submit a build request for the package by clicking on Build, and the Build Request Form will be displayed. Select the build Environment, then click the Submit button to initiate the build execution.

![5  Build](https://github.com/FlexDeploy-Community/community/assets/125540444/2eeb6bb7-7cda-4e5a-be8b-c3948c1dac22)

**2. Review Build Execution Details:**
Following a successful Build operation, here you can take note of the build execution details for your package. For more detailed information, click on the generated Execution Id. Here, from Artifact option you can download data (.zip) as per your package files setup.

![6  Artifact](https://github.com/FlexDeploy-Community/community/assets/125540444/1eabc094-ca44-4c04-b462-ef6fda1a76f2)

**3. Initiate Deployment:**
Begin the deployment process by clicking on "Submit Deploy Request". This action opens the Deploy Request Form. Select the target environment where you intend to deploy the package and click the "Submit" button. This formally submits the deployment request for the specified package.

![7  Deploy](https://github.com/FlexDeploy-Community/community/assets/125540444/ee135c6d-5b69-4ea8-90bf-5ba9eec6da6e)

**4. Review Deployment Execution:**
Following a successful deployment operation, click on the generated Execution ID to access more detailed information about the deployment process.

**5. Check Data Deployment Reports:**
Within the Reports option, you can examine data deployment reports that are generated after every deployment operation, whether it was successful or encountered failures. In the case of a failed deployment, the report provides specific details about the reasons for the failure.

![8  Reports](https://github.com/FlexDeploy-Community/community/assets/125540444/22698f07-d2b4-4578-92d5-13a8a8d2725b)

**6. Verify Data Deployment:**
Ensure that the data and records for all the listed objects have been effectively deployed to the target environment.

By following these sequential steps, FlexDeploy streamlines and automates the data deployment lifecycle, providing organizations with a robust and efficient solution for managing Salesforce data across various environments.

**Comment by davidwilson12345** [comment]
Great post on Salesforce deployment! Deployment strategies are crucial for ensuring smooth and efficient implementation of Salesforce solutions. If anyone is looking to streamline their Salesforce integration process, check out our product  [200 OK](https://sforce.co/3TtBkk4) for expert guidance and solutions tailored to your needs. Don't miss out on optimizing your Salesforce experience!
